wagtail/admin/views/reports/page_types_usage.py

```diff
@@ -0,0 +1,191 @@
+import django_filters
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+from django.db.models import Count, OuterRef, Q, Subquery
+from django.utils.translation import gettext_lazy as _
+
+from wagtail.admin.filters import WagtailFilterSet
+from wagtail.admin.views.reports import ReportView
+from wagtail.coreutils import get_content_languages
+from wagtail.models import ContentType, Page, PageLogEntry, Site, get_page_models
+from wagtail.users.utils import get_deleted_user_display_name
+
+
+def _get_locale_choices():
+    choices = [
+        (language_code, display_name)
+        for language_code, display_name in get_content_languages().items()
+    ]
+    return choices
+
+
+def _get_site_choices():
+    choices = [(site.pk, str(site)) for site in Site.objects.all()]
+    return choices
+
+
+def _annotate_last_edit_info(queryset):
+    latest_edit_log = PageLogEntry.objects.filter(content_type=OuterRef("pk")).order_by(
+        "-timestamp", "-pk"
+    )
+
+    queryset = queryset.annotate(
+        count=Count("pages"),
+        last_edited_page=Subquery(latest_edit_log.values("page")[:1]),
+        last_edited_by=Subquery(latest_edit_log.values("user")[:1]),
+    )
+
+    return queryset
+
+
+def _annotate_last_edit_info_by_locale(queryset, language_code):
```

Comment:
This function and the one above are quite similar, really. Maybe we could bring them together in a `annotate_queryset(queryset, form)` method on our FilterSet class, and add conditions for the conditional bits?

Reply:
Do these annotations also need to take the selected `site` into account? I guess that's the sort of thing that would be easier with a consolidated method?

Reply:
Sure, I can bring them together in one function.

The `site` is taken into account before we annotate. On `filter_queryset`'s method of `PageTypesUsageReportFilterSet`, the `ContentType`'s are filtered on a `site` such that pages from that `ContentType` have the same root as the `site`'s root.

When we get to do the annotations, the pages have already been filtered. As I type this, I'm questioning myself if that's actually true :laughing: I'll add a test to confirm that and if there is a need to filter the annotations I can do that from a consolidated method.

Reply:
I've added the site filter into the annotations as well, as they were indeed needed.